copyrightify/copyrightify.py

```diff
@@ -0,0 +1,163 @@
+            raise RuntimeError("Please run `git config --global user.{0} '<{0}>'``".format(key))
+
+    for entry in ['name', 'email', 'company']:
+        context[entry] = get_user_config(entry)
+    context['year'] = datetime.datetime.now().year,
+
+    return context
+
+
+def find_shebang(line, filetypes):
+    """Search line for a shebang, and attempt to extract file type.
+    """
+    shebang = None
+    filetype = None
+    if SHEBANG_REGEX.match(line):
+        shebang = line
+        filetypes = list(filetypes.keys())
+        filetypes.sort(key=len, reverse=True)
+        for extension in filetypes:
+            # This is a little hacky, but should detect any interpreters we currently use (python, bash). May break.
+            if extension in shebang:
+                filetype = extension
+                break
+
+    return shebang, filetype
+
+
+def process_file(path, config, context):
+    """Determine file type (whether via shebang or file extension), and apply header if necessary.
+    """
+    filetypes = config['filetypes']
+    preamble = config['preamble']
+    license_text = config['licenses'][context['license']]
+
+    with open(path) as f:
+        try:
+            shebang_candidate = f.readline()
+        except UnicodeDecodeError:
+            # Not a unicode text file, skipping
+            return
+
+        shebang, filetype = find_shebang(shebang_candidate, filetypes)
+        f.seek(0)
+
+        if not filetype:  # Only use file extension if shebang didn't help
+            filetype = os.path.splitext(path)[1].lstrip('.')
+
+        try:
+            comment_start, line_prefix, comment_end = filetypes[filetype]
+        except KeyError:
+            print("Extension is unsupported, skipping {}".format(path))
+            return
+
+        template = str()
+
+        if comment_start:
+            template += comment_start + '\n'
+
+        for line in (preamble + '\n' + license_text).splitlines():
+            template += (line_prefix + line).rstrip(' ') + '\n'
+
+        template += comment_end
+
+        header = jinja2.Environment().from_string(template).render(
+            filename=os.path.split(path)[1],
+            **context)
+
+        # Make a temporary file that will replace the original source
+        with tempfile.NamedTemporaryFile(mode='w') as tmp:
+            if shebang:
+                tmp.write(shebang)  # Make sure shebang is preserved as first line
+
+            print(header, file=tmp)
+
+            for idx, line in enumerate(f):
+                if idx == 0 and shebang:
+                    continue  # skip copying shebang if we detected one earlier
+
+                if idx < COPYRIGHT_SCAN_LINES and 'Copyright' in line:
+                    print("Contains copyright already, skipping {}".format(path))
+                    return
+                tmp.write(line)
+
+            tmp.flush()
+            print("Added copyright to {}".format(path))
+            shutil.copyfile(tmp.name, path)
```

Comment:
This can fail, e.g. due to file permissions. At minimum, might be good to move the print after it.